src/snowflake/snowpark/modin/pandas/dataframe.py

```diff
@@ -153,53 +152,80 @@ def __init__(
+                    [0] if columns is None else columns
+                )
+            if columns is not None and data.name not in columns:
                 # If the columns provided are not in the named Series, pandas clears
                 # the DataFrame and sets columns to the columns provided.
-                elif columns is not None and data.name not in columns:
-                    self._query_compiler = from_pandas(
-                        self.__constructor__(columns=columns)
-                    )._query_compiler
-                if index is not None:
-                    self._query_compiler = data.loc[index]._query_compiler
-            elif columns is None and index is None:
+                query_compiler = from_pandas(
+                    self.__constructor__(columns=columns)
+                )._query_compiler  # pragma: no cover
+
+        elif isinstance(data, DataFrame):
+            # CASE 5: data is a Snowpark pandas DataFrame
+            query_compiler = data._query_compiler.copy()
+
+            if columns is None and index is None:
+                # If the new DataFrame has the same columns and index as the original DataFrame,
+                # the query compiler is shared and kept track of as a sibling.
+                self._query_compiler = query_compiler
                 data._add_sibling(self)
-            else:
-                if columns is not None and any(i not in data.columns for i in columns):
-                    ErrorMessage.not_implemented(
-                        "Passing non-existant columns or index values to constructor not"
-                        + " yet implemented."
-                    )  # pragma: no cover
-                if index is None:
-                    index = slice(None)
-                if columns is None:
-                    columns = slice(None)
-                self._query_compiler = data.loc[index, columns]._query_compiler
-
-        # Check type of data and use appropriate constructor
-        elif query_compiler is None:
-            distributed_frame = from_non_pandas(data, index, columns, dtype)
-            if distributed_frame is not None:
-                self._query_compiler = distributed_frame._query_compiler
                 return
+            # The `columns` parameter is used to select the columns from `data` that will be in the resultant
+            # DataFrame. If a value in `columns` is not present in `data`'s columns, it will be added as a
+            # new column filled with NaN values. These columns are tracked by the `extra_columns` variable.
+            if data.columns is not None and columns is not None:
+                extra_columns = [col for col in columns if col not in data.columns]
+            else:
+                extra_columns = []
+            query_compiler = data._query_compiler.create_qc_with_extra_columns(
+                extra_columns
+            )
 
+        else:
+            # CASE 5: Non-Snowpark pandas data
```

Comment:
why not just call from_pandas for all cases?
What is case 5.A?

Reply:
I believe the only special case is 
Special case: data is a dictionary where all the values are Snowpark pandas Series

Reply:
I refactored it all - the two special cases are dict or list with all snowpark pandas elements